src/Type/Php/RegexArrayShapeMatcher.php

```diff
@@ -484,4 +500,49 @@ private function walkRegexAst(
 		}
 	}
 
+	private function getPatternType(Expr $patternExpr, Scope $scope): Type
+	{
+		if ($patternExpr instanceof Expr\BinaryOp\Concat) {
+			return $this->resolvePatternConcat($patternExpr, $scope);
```

Comment:
I don't get it, what's happening here? Shouldn't this be done instead by improving `MutatingScope::resolveType`? By introducing dynamic return type extension for preg_quote or something?

Reply:
I'll try to explain as I feel this is my fault :D

I think you can't reasonably treat preg quote's return value as a constant string (empty or "x") in any other context than here when trying to get a constant pattern string so that you have a regex to parse.

As such it feels appropriate to me to hack this in here but maybe there are better ways 🤷🏻‍♂️

Reply:
Its pretty unlikely that we know the args to preg_quote() at static analysis time in a real world use-case. 

Please also look at the added test-cases. These wouldn't work with a return type extension, as the extension couldn't produce a constant-string for these cases. 

So while adding a preg_quote return type extension could make sense in the future, it would not help with the examples at hand. The ArrayShapeMatcher just can ignore the preg_quote() parts and still analyse the shape, as long as all other parts of a concatenation are constant

Reply:
Alright, in that case the code needs some comments. I did not understand at all why you are returning an empty string for a preg_quote call.

Reply:
done 👍